Common/ModelInfo.swift

```diff
@@ -0,0 +1,295 @@
+//
+//  ModelInfo.swift
+//  Diffusion
+//
+//  Created by Pedro Cuenca on 29/12/22.
+//  See LICENSE at https://github.com/huggingface/swift-coreml-diffusers/LICENSE
+//
+
+import CoreML
+
+enum AttentionVariant: String {
+    case original
+    case splitEinsum
+    case splitEinsumV2
+}
+
+extension AttentionVariant {
+    var defaultComputeUnits: MLComputeUnits { self == .original ? .cpuAndGPU : .cpuAndNeuralEngine }
+}
+
+/// Track if the model is downloaded and on the filesystem, ready for use.
+enum ModelReadinessState {
+    case unknown
+    case downloading
+    case downloaded
+    case uncompressing
+    case ready
+    case failed
+}
+
+class ModelReadiness: ObservableObject {
+    let modelInfo: ModelInfo
+    @Published var state: ModelReadinessState
+    
+    init(modelInfo: ModelInfo, state: ModelReadinessState) {
+        self.modelInfo = modelInfo
+        self.state = state
+    }
+}
+
+struct ModelInfo: Hashable, Identifiable {
+    
+    // To be Identifiable compliant we need an id: field.
+    let id = UUID()
+
+    /// Hugging Face model Id that contains .zip archives with compiled Core ML models
+    let modelId: String
+    
+    /// Arbitrary string for presentation purposes. Something like "2.1-base"
+    let modelVersion: String
+    
+    /// Which attention variant is this model associated with?
+    let variant: AttentionVariant
```

Comment:
The reason the `variant` was not part of the model abstraction is because `ModelInfo` instances are presented in the UI as single entities. If you select a different attention variant, it's still the same model. If we go this way, then we'd have duplicate entries in the model selector (one for `original`, another one for `split_einsum`) for each model; or we'd need to complicate the UI logic.

(Arguably, the `palettized` versions could have been handled in exactly the same way (a single model entry would have several variants such as `original`, `original palettized`, `split_einsum`, `split_einsum_v2 palettized`, ...), but I separated them because the palettized versions are two new and different, and they only work in beta versions of the OS).

Reply:
I am not a user of the hugging face API and probably have some misunderstandings on how the model and variant combinations work. The reason for separating the ModelInfo into a discrete instance for each variant was due to my user experience.

If I wanted to download a third party model from HF via the web I would go to 
https://huggingface.co/coreml/coreml-GTA5_Artwork_Diffusion
There I would find in Files and Versions a split_einsum and an original model. After downloading and extracting these two matched models I would have one folder "gta5ArtworkDiffusion-v1_ema-vae_original_compiled" and another folder "GTA5_Artwork_Diffusion_split-einsum_compiled." 

When I deconstruct these two folder names I get one with a bidding version and the two mismatching on their name due to a spacing issue (and a caps issue which is minor.)  

How can I programatically look through a models folder in an unknown state that over time has had lots of slightly incorrectly named models landed into it? 

Since I found a few of these that would result in multiple models anyway I started down the path of treating each folder as its own ModelInfo unit.

Further the reason for the file monitor is that whenever the folder contents change there is a notification which updates the UI. These two are connected. Let me show you how and maybe then you can update my understanding...

Let's assume a file list of:
elldrethsVividMix-v2-VividEr_original
analogDream3D_1_ema_vae-split_einsum_compiled
coreml-stable-diffusion-1-4_original_compiled
coreml-stable-diffusion-2-1-base_split_einsum_compiled
coreml-stable-diffusion-v1-5_original_compiled
Deliberate_v2_original
deliberate_v2_split-einsum

<img width="286" alt="Screenshot 2023-06-22 at 16 54 13" src="https://github.com/huggingface/swift-coreml-diffusers/assets/47501428/73bd12ec-baf7-4e1a-b319-faae8ea8ed4c">

With CPU and GPU on a Mac you get a file list of 
CoreML Stable Diffusion v1.4
CoreML Stable Diffusion v1.5
Deliberate
elderethsVi [note: no idea why this one truncates!]

When I switch to All compute units on Mac I get 

<img width="286" alt="Screenshot 2023-06-22 at 16 56 29" src="https://github.com/huggingface/swift-coreml-diffusers/assets/47501428/948065cd-ae6a-4fec-9722-8bf986a7d611">
CoreML Stable Diffusion 2 1 base
Deliberate
analogDream3d

Because a filtered list of model info objects are made that align their variant support. Therefore when the UI compute units are updated and that affects which variants should be used the UI refreshes and the list of available models updates.

Does this attempt at filtering the available selections make sense or is it a misunderstanding of the user case for changing the compute units manually?

Reply:
I see, it's unfortunate that we have these naming inconsistencies.

My initial assumption was that models were always available in the two variants, so the list was independent of the compute units selected. However, if you select a model variant that had not been downloaded yet, the download was triggered.

In your approach the compute units act as a filter. It's not a bad solution for this use case, where we no longer have the guarantee that all variants exist or share the same name. There could be some slightly confusing situations, say you run the app with `Deliberate_v2_original` selected (and `deliberate_v2_split-einsum` not downloaded), then select `CPU and Neural Engine`: the `Deliberate` model will disappear from the list. I think this can be disconcerting for first-time users that don't really know the details of how everything works. Not sure how to deal with it though.

Reply:
The long term ideal solution is to have an integrated HF API browser. If the app displays and downloads assets directly then the original approach trying to access a missing file from the hub triggering a known download link makes sense. 

However, I found that there are models listed in GitHub which are not submitted to the HF website so I would still suspect the user might download a completely unknown folder with a completely unknown naming structure (or rename it themselves for some reason, or maybe work with their own test models.)

Therefore I think it's best solve for an unknown state local model folder. Even after later work results in a more elegant user experience with a HF browser UI it'll still be useful for the intermediate or advanced users.

The instance where Deliberate_v2_original is found and deliberate_v2_split_einsum is missing would persist and we still wouldn't know how to find the matching file to help the user. The best solution I could envision at the moment is to add a hint in the UI. Instead of the Disclosure section being named "Hub Models" we rename it to reflect the selected filter as it changes "Models (original)" "Models (einsum)" or something to alert the intermediate/advanced user which file they're missing on their system.